data_processing/create_heatmap.py

```diff
@@ -141,21 +92,74 @@ def create_width_dict():
         "virus-like-particle": 79.07
     }
 
-def process_tomogram(json_folder, image_shape, eps=0.00001, sigma=None, lower_bound=None, upper_bound=None):
+def create_heatmap(json_folder, image_shape, eps=0.00001, sigma=None, lower_bound=None, upper_bound=None, bb=None):
     """
-    Process a tomogram by creating a heatmap based on protein data from JSON files.
+    Process a tomogram by creating a heatmap based on protein data from JSON files,
+    with optional bounding box constraints.
 
     Parameters:
         json_folder (str): Path to the folder containing JSON files.
         image_shape (tuple): Shape of the 3D image (z, y, x).
+        eps (float): Threshold for truncating the Gaussian.
+        sigma (float, optional): Fixed sigma value for Gaussian.
+        lower_bound (float, optional): Minimum allowed sigma.
+        upper_bound (float, optional): Maximum allowed sigma.
+        bb (tuple, optional): Bounding box (z_min, z_max, y_min, y_max, x_min, x_max).
 
     Returns:
         np.ndarray: Generated 3D heatmap.
     """
-    json_files = [os.path.join(json_folder, f) for f in os.listdir(json_folder) if f.endswith('.json')]
+    picks_folder = os.path.join(json_folder, "Picks")
+    json_files = [os.path.join(picks_folder, f) for f in os.listdir(picks_folder) if f.endswith('.json')]
     coordinates, protein_types = parse_json_files(json_files)
     width_dict = create_width_dict()
-    return create_heatmap(image_shape, coordinates, protein_types, width_dict, eps, sigma, lower_bound, upper_bound)
+    
+    if bb:
+        (z_min, z_max), (y_min, y_max), (x_min, x_max) = [(s.start, s.stop) for s in bb]
+        restricted_shape = (z_max - z_min, y_max - y_min, x_max - x_min)
+        heatmap = np.zeros(restricted_shape)
+    else:
+        heatmap = np.zeros(image_shape)
+    
+    for coord, protein in zip(coordinates, protein_types):
+        z, y, x = map(int, coord)
+        width = width_dict.get(protein, 1.0)
+        gaussian = create_gaussian_stamp(int(width * 0.3), eps, lower_bound, upper_bound)
```

Comment:
This only depends on the protein type, so you can pre-compute this for all protein types, store it in a dict, and then just access the gaussian for this dict here. This will make everything much more efficient. (You only have to compute this 6 times instead of for each coordinate)